File: src/main/java/infinitespire/relics/Cupcake.java
package infinitespire.relics;

import com.megacrit.cardcrawl.actions.common.ApplyPowerAction;
import com.megacrit.cardcrawl.core.EnergyManager;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import infinitespire.InfiniteSpire;
import infinitespire.abstracts.Relic;
import infinitespire.powers.DeenergizedPower;

public class Cupcake extends Relic {
	public static final String ID = InfiniteSpire.createID("Cupcake");
	public static final String NAME = "Cupcake";
	public boolean firstTurn = false;
	public boolean tryToAddPower = false;
	
	public Cupcake() {
		super(ID, "cupcake", RelicTier.BOSS, LandingSound.MAGICAL);
		this.counter = 3;
	}

	@Override
	public void onEquip() {
		EnergyManager energy = AbstractDungeon.player.energy;
		++energy.energyMaster;
	}

	@Override
	public void onUnequip() {
		EnergyManager energy = AbstractDungeon.player.energy;
		--energy.energyMaster;
	}

	@Override
	public void atBattleStart() {
		counter = 3;
		tryToAddPower = true;
	}

	@Override
	public void onVictory() {
		counter = 3;
	}

	@Override
	public void onPlayerEndTurn() {
		if(counter > 0) counter --;
		if(counter == 0){
			if(!AbstractDungeon.player.hasPower(DeenergizedPower.powerID) && tryToAddPower) {
				AbstractDungeon.actionManager.addToBottom(new ApplyPowerAction(
					AbstractDungeon.player,
					AbstractDungeon.player,
					new DeenergizedPower(AbstractDungeon.player, 1),
					1));
			}
			tryToAddPower = false;
		}
	}
}

File: src/main/kotlin/com/blanktheevil/infinitespire/relics/Cupcake.kt
package com.blanktheevil.infinitespire.relics

import com.blanktheevil.infinitespire.extensions.applyPower
import com.blanktheevil.infinitespire.extensions.makeID
import com.blanktheevil.infinitespire.extensions.player
import com.blanktheevil.infinitespire.powers.DeEnergizedPower

class Cupcake : Relic(ID, IMG, TIER, SOUND) {
  companion object {
    val ID = "Cupcake".makeID()
    private const val IMG = "cupcake"
    private val TIER = RelicTier.BOSS
    private val SOUND = LandingSound.MAGICAL
  }

  private var tryToAddPower = false

  init {
    counter = 3
  }

  override fun onEquip() {
    player.energy.energyMaster++
  }

  override fun onUnequip() {
    player.energy.energyMaster--
  }

  override fun atBattleStart() {
    counter = 3
    tryToAddPower = true
  }

  override fun onVictory() {
    counter = 3
  }

  override fun onPlayerEndTurn() {
    if (counter > 0) counter--
    if (counter == 0) {
      if (!player.hasPower(DeEnergizedPower.powerID) && tryToAddPower) {
        player.applyPower(
          DeEnergizedPower(player, 1),
          amount = 1
        )
      }
      tryToAddPower = false
    }
  }
}